src/rgw/rgw_auth.h

```diff
@@ -51,4 +58,243 @@ inline ostream& operator<<(ostream& out, const RGWIdentityApplier &id) {
+                                       const AuthInfo info) const = 0;
+  };
+};
+
+
+/* Local auth applier targets those auth engines that store user information
+ * in the RADOS store. As a consequence of performing the authentication, they
+ * will have the RGWUserInfo structure loaded. Exploiting this trait allows to
+ * avoid additional call to underlying RADOS store. */
+class RGWLocalAuthApplier : public RGWAuthApplier {
+protected:
+  const RGWUserInfo user_info;
+  const std::string subuser;
+
+  uint32_t get_perm_mask(const std::string& subuser_name,
+                         const RGWUserInfo &uinfo) const;
+
+public:
+  static const std::string NO_SUBUSER;
+
+  RGWLocalAuthApplier(CephContext * const cct,
+                      const RGWUserInfo& user_info,
+                      const std::string subuser)
+    : RGWAuthApplier(cct),
+      user_info(user_info),
+      subuser(subuser) {
+  }
+
+
+  virtual int get_perms_from_aclspec(const aclspec_t& aclspec) const override;
+  virtual bool is_admin_of(const rgw_user& uid) const override;
+  virtual bool is_owner_of(const rgw_user& uid) const override;
+  virtual int get_perm_mask() const override {
+    return get_perm_mask(subuser, user_info);
+  }
+  virtual void load_acct_info(RGWUserInfo& user_info) const override; /* out */
+
+  struct Factory {
+    virtual ~Factory() {}
+    virtual aplptr_t create_apl_local(CephContext * const cct,
+                                      const RGWUserInfo& user_info,
+                                      const std::string& subuser) const = 0;
+    };
+};
+
+
+/* Abstract class for authentication backends (auth engines) in RadosGW.
+ *
+ * An engine is supposed only to:
+ *  - authenticate (not authorize!) a given request basing on req_state,
+ *  - provide an upper layer with RGWAuthApplier to commit all changes to
+ *    data structures (like req_state) and to the RADOS store (creating
+ *    an account, synchronizing user personal info).
+ *    Auth engine MUST NOT make any changes to req_state nor RADOS store.
+ *
+ * Separation between authentication and global state modification has been
+ * introduced because many auth engines are perfectly orthogonal to applier
+ * and thus they can be decoupled. Additional motivation is clearly distinguish
+ * all places which can modify underlying data structures. */
+class RGWAuthEngine {
+protected:
+  CephContext * const cct;
+
+  RGWAuthEngine(CephContext * const cct)
+    : cct(cct) {
+  }
+  /* Make the engines non-copyable and non-moveable due to const-correctness
+   * and aggregating applier factories less costly and error-prone. */
+  RGWAuthEngine(const RGWAuthEngine&) = delete;
+  RGWAuthEngine& operator=(const RGWAuthEngine&) = delete;
+
+public:
+  /* Fast, non-throwing method for screening whether a concrete engine may
+   * be interested in handling a specific request. */
+  virtual bool is_applicable() const noexcept = 0;
```

Comment:
All necessary info should be taken and stored by a constructor of a derived class to guarantee that all methods will work on the same context.